### ISSUE TITLE
$& value does not properly propagate into replacement body of s///

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7120]
Migrated from [rt.perl.org#26986](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=26986) (status was 'resolved')

Searchable as RT26986$

**Comment by p5pRT** [comment]
### From perl-5.8.0@ton.iguana.be

#### Created by perl-5.8.0@ton.iguana.be



perl \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./\<$F\[@&#8203;a="b"\.\.$&\]>/; print; print"a=@&#8203;a"'
Argument "b" isn't numeric in range \(or flop\) at \-e line 1\.
Use of uninitialized value in range \(or flop\) at \-e line 1\.
Use of uninitialized value in range \(or flop\) at \-e line 1\.
\<B>
a=0

I expected $& to be "a"\, so @&#8203;a should become \(a b c \.\.\.\.z\)\, so fetch 
element 26 from @&#8203;F\, which should be "Z"

The 'Argument "b" isn't numeric in range \(or flop\) at \-e line 1\.' seems to
indicate $& is undef which in my perl still causes @&#8203;a to become \(0\)
\(this was bug 24735\, but even with that fixed I just expect that now
@&#8203;a will be \(\) and it will output \<A>\, which is still wrong\)\.

However\, mentioning $& in the real string body will make things work 
as expected&#8203;:
perl \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./$&\<$F\[@&#8203;a="b"\.\.$&\]>/; print; print"a=@&#8203;a"'
a\<Z>
a=b c d e f g h i j k l m n o p q r s t u v w x y z

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.8.2:

Configured by ton at Sun Jan  4 19:19:06 CET 2004.

Summary of my perl5 (revision 5.0 version 8 subversion 2) configuration:
  Platform:
    osname=linux, osvers=2.6.0, archname=i686-linux-64int-ld
    uname='linux quasar 2.6.0 #3 thu dec 18 18:22:48 cet 2003 i686 gnulinux '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=define use64bitall=undef uselongdouble=define
    usemymalloc=y, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -fomit-frame-pointer',
    cppflags='-fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='3.4.0 20031231 (experimental)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long long', ivsize=8, nvtype='long double', nvsize=12, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.3.2.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.3.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.2:
    /usr/lib/perl5/5.8.2/i686-linux-64int-ld
    /usr/lib/perl5/5.8.2
    /usr/lib/perl5/site_perl/5.8.2/i686-linux-64int-ld
    /usr/lib/perl5/site_perl/5.8.2
    /usr/lib/perl5/site_perl
    .


Environment for perl v5.8.2:
    HOME=/home/ton
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/ton/bin.Linux:/home/ton/bin:/home/ton/bin.SampleSetup:/usr/local/bin:/usr/local/sbin:/usr/local/jre/bin:/home/oracle/product/9.0.1/bin:/usr/local/ar/bin:/usr/games/bin:/usr/X11R6/bin:/usr/share/bin:/usr/bin:/usr/sbin:/bin:/sbin:.
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @cwest

It was Monday\, February 23\, 2004 when perl\-5\.8\.0@&#8203;ton\.iguana\.be \(via RT\) took the soap box\, saying&#8203;:
: perl \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./\<$F\[@&#8203;a="b"\.\.$&\]>/; print; print"a=@&#8203;a"'
: Argument "b" isn't numeric in range \(or flop\) at \-e line 1\.
: Use of uninitialized value in range \(or flop\) at \-e line 1\.
: Use of uninitialized value in range \(or flop\) at \-e line 1\.
: \<B>
: a=0


If I add a /e modifier on the end of your s///\, and fix up the
replacement section\, I get the following&#8203;:

perl \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./join $F\[@&#8203;a="b"\.\.$&\]\, "\<"\, ">"/e; print; print"a=@&#8203;a"'
\<Z>
a=b c d e f g h i j k l m n o p q r s t u v w x y z

: I expected $& to be "a"\, so @&#8203;a should become \(a b c \.\.\.\.z\)\, so fetch 
: element 26 from @&#8203;F\, which should be "Z"

I think that I got what you expected\. When you want to evaluate code
in a regular expression you must add the /e modifier\. I think that
this is notabug\. Am I missing something?

&nbsp;  Casey West

\-\- 
Good Idea&#8203;: Taking a deep breath before jumping into a swimming pool\.
Bad Idea&#8203;:  Taking a deep breath after jumping into a swimming pool\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @chipdude

According to Casey West&#8203;:
> It was Monday\, February 23\, 2004 when perl\-5\.8\.0@&#8203;ton\.iguana\.be \(via RT\) took the soap box\, saying&#8203;:
> : perl \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./\<$F\[@&#8203;a="b"\.\.$&\]>/; print; print"a=@&#8203;a"'
> : Argument "b" isn't numeric in range \(or flop\) at \-e line 1\.
> : Use of uninitialized value in range \(or flop\) at \-e line 1\.
> : Use of uninitialized value in range \(or flop\) at \-e line 1\.
> : \<B>
> : a=0
> 
> If I add a /e modifier on the end of your s/// \.\.\.

Oh\, that's interesting\.  I think it's a real bug\.

The subscript of an array element is always evaluated as code\, even in
a double\-quoted string&#8203;:

&nbsp;    $ perl \-le '@&#8203;F=qw\(a b c d e\); print "$F\[1\+2\]"'
&nbsp;    d

I see no reason why C\<1\+2> should work but C\<@&#8203;a = "b" \.\. $&> should not\.
\-\- 
Chip Salzenberg               \- a\.k\.a\. \-               \<chip@&#8203;pobox\.com>
"I wanted to play hopscotch with the impenetrable mystery of existence\,
&nbsp;    but he stepped in a wormhole and had to go in early\."  // MST3K

**Comment by p5pRT** [comment]
### From @tamias

On Tue\, Feb 24\, 2004 at 11&#8203;:56&#8203;:51AM \-0500\, Casey West wrote&#8203;:
> It was Monday\, February 23\, 2004 when perl\-5\.8\.0@&#8203;ton\.iguana\.be \(via RT\) took the soap box\, saying&#8203;:
> : perl \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./\<$F\[@&#8203;a="b"\.\.$&\]>/; print; print"a=@&#8203;a"'
> : Argument "b" isn't numeric in range \(or flop\) at \-e line 1\.
> : Use of uninitialized value in range \(or flop\) at \-e line 1\.
> : Use of uninitialized value in range \(or flop\) at \-e line 1\.
> : \<B>
> : a=0
> 
> If I add a /e modifier on the end of your s///\, and fix up the
> replacement section\, I get the following&#8203;:
> 
> perl \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./join $F\[@&#8203;a="b"\.\.$&\]\, "\<"\, ">"/e; print; print"a=@&#8203;a"'
> \<Z>
> a=b c d e f g h i j k l m n o p q r s t u v w x y z
> 
> : I expected $& to be "a"\, so @&#8203;a should become \(a b c \.\.\.\.z\)\, so fetch 
> : element 26 from @&#8203;F\, which should be "Z"
> 
> I think that I got what you expected\. When you want to evaluate code
> in a regular expression you must add the /e modifier\. I think that
> this is notabug\. Am I missing something?

Perl evaluates code inside array and hash indexes inside double\-quotish
strings\.  MJD takes advantage of this with his Interpolation module
\(e\.g\. print "3 \+ 4 = $E\{3\+4\}";\)\.  The warnings from Ton's example indicate
that the flip flop operator is being executed; the bug is that $& isn't
being set\.

Ronald

**Comment by p5pRT** [comment]
### From @Hugmeir

On Tue Feb 24 14&#8203;:12&#8203;:49 2004\, rjk\-perl\-p5p@&#8203;tamias\.net wrote&#8203;:
> On Tue\, Feb 24\, 2004 at 11&#8203;:56&#8203;:51AM \-0500\, Casey West wrote&#8203;:
> > It was Monday\, February 23\, 2004 when perl\-5\.8\.0@&#8203;ton\.iguana\.be \(via
> RT\) took the soap box\, saying&#8203;:
> > : perl \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./\<$F\[@&#8203;a="b"\.\.$&\]>/; print;
> print"a=@&#8203;a"'
> > : Argument "b" isn't numeric in range \(or flop\) at \-e line 1\.
> > : Use of uninitialized value in range \(or flop\) at \-e line 1\.
> > : Use of uninitialized value in range \(or flop\) at \-e line 1\.
> > : \<B>
> > : a=0
> >
> > If I add a /e modifier on the end of your s///\, and fix up the
> > replacement section\, I get the following&#8203;:
> >
> > perl \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./join $F\[@&#8203;a="b"\.\.$&\]\, "\<"\,
> ">"/e; print; print"a=@&#8203;a"'
> > \<Z>
> > a=b c d e f g h i j k l m n o p q r s t u v w x y z
> >
> > : I expected $& to be "a"\, so @&#8203;a should become \(a b c \.\.\.\.z\)\, so
> fetch
> > : element 26 from @&#8203;F\, which should be "Z"
> >
> > I think that I got what you expected\. When you want to evaluate code
> > in a regular expression you must add the /e modifier\. I think that
> > this is notabug\. Am I missing something?
> 
> Perl evaluates code inside array and hash indexes inside double\-
> quotish
> strings\.  MJD takes advantage of this with his Interpolation module
> \(e\.g\. print "3 \+ 4 = $E\{3\+4\}";\)\.  The warnings from Ton's example
> indicate
> that the flip flop operator is being executed; the bug is that $&
> isn't
> being set\.
> 
> Ronald
> 

This is still present in blead\, albeit behaving a tad differently&#8203;:
$ \./perl \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./\<$F\[@&#8203;a="b"\.\.$&\]>/; print;
print"a=@&#8203;a"'
Use of uninitialized value $& in range \(or flop\) at \-e line 1\.
\<A>
a=

Strangely enough\, you can get $& to work in the flipflop if it's
evaluated beforehand somehow&#8203;:
$ \./perl \-Ilib \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./\<$F\[\!\!$& &&
\(@&#8203;a="b"\.\.$&\)\]>/; print; print"a=@&#8203;a"'
\<Z>
a=b c d e f g h i j k l m n o p q r s t u v w x y z

Putting an sv\_dump\(\) after the SvGETMAGIC\(right\) in pp\_flop gives us this&#8203;:
$ \./perl \-Ilib \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./\<$F\[@&#8203;a="b"\.\.$&\]>/; print;
print"a=@&#8203;a"'
\[\.\.\.\]
SV = PVMG\(0x854ab74\) at 0x855c02c
&nbsp;  REFCNT = 1
&nbsp;  FLAGS = \(GMG\,SMG\)
&nbsp;  IV = 0
&nbsp;  NV = 0
&nbsp;  PV = 0
&nbsp;  MAGIC = 0x855d3f4
&nbsp;    MG\_VIRTUAL = &PL\_vtbl\_sv
&nbsp;    MG\_TYPE = PERL\_MAGIC\_sv\(\\0\)
&nbsp;    MG\_OBJ = 0x855c01c
&nbsp;    MG\_LEN = 1
&nbsp;    MG\_PTR = 0x855d314 "&"
Use of uninitialized value $& in range \(or flop\) at \-e line 1\.
\<A>
a=


I don't know the first thing about magic or the regex engine\, so I'm
stopping here\. Could someone take a peek?

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Apr 29 09&#8203;:45&#8203;:23 2012\, Hugmeir wrote&#8203;:
> On Tue Feb 24 14&#8203;:12&#8203;:49 2004\, rjk\-perl\-p5p@&#8203;tamias\.net wrote&#8203;:
> > On Tue\, Feb 24\, 2004 at 11&#8203;:56&#8203;:51AM \-0500\, Casey West wrote&#8203;:
> > > It was Monday\, February 23\, 2004 when perl\-5\.8\.0@&#8203;ton\.iguana\.be \(via
> > RT\) took the soap box\, saying&#8203;:
> > > : perl \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./\<$F\[@&#8203;a="b"\.\.$&\]>/; print;
> > print"a=@&#8203;a"'
> > > : Argument "b" isn't numeric in range \(or flop\) at \-e line 1\.
> > > : Use of uninitialized value in range \(or flop\) at \-e line 1\.
> > > : Use of uninitialized value in range \(or flop\) at \-e line 1\.
> > > : \<B>
> > > : a=0
> > >
> > > If I add a /e modifier on the end of your s///\, and fix up the
> > > replacement section\, I get the following&#8203;:
> > >
> > > perl \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./join $F\[@&#8203;a="b"\.\.$&\]\, "\<"\,
> > ">"/e; print; print"a=@&#8203;a"'
> > > \<Z>
> > > a=b c d e f g h i j k l m n o p q r s t u v w x y z
> > >
> > > : I expected $& to be "a"\, so @&#8203;a should become \(a b c \.\.\.\.z\)\, so
> > fetch
> > > : element 26 from @&#8203;F\, which should be "Z"
> > >
> > > I think that I got what you expected\. When you want to evaluate code
> > > in a regular expression you must add the /e modifier\. I think that
> > > this is notabug\. Am I missing something?
> > 
> > Perl evaluates code inside array and hash indexes inside double\-
> > quotish
> > strings\.  MJD takes advantage of this with his Interpolation module
> > \(e\.g\. print "3 \+ 4 = $E\{3\+4\}";\)\.  The warnings from Ton's example
> > indicate
> > that the flip flop operator is being executed; the bug is that $&
> > isn't
> > being set\.
> > 
> > Ronald
> > 
> 
> This is still present in blead\, albeit behaving a tad differently&#8203;:
> $ \./perl \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./\<$F\[@&#8203;a="b"\.\.$&\]>/; print;
> print"a=@&#8203;a"'
> Use of uninitialized value $& in range \(or flop\) at \-e line 1\.
> \<A>
> a=
> 
> Strangely enough\, you can get $& to work in the flipflop if it's
> evaluated beforehand somehow&#8203;:
> $ \./perl \-Ilib \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./\<$F\[\!\!$& &&
> \(@&#8203;a="b"\.\.$&\)\]>/; print; print"a=@&#8203;a"'
> \<Z>
> a=b c d e f g h i j k l m n o p q r s t u v w x y z
> 
> Putting an sv\_dump\(\) after the SvGETMAGIC\(right\) in pp\_flop gives us this&#8203;:
> $ \./perl \-Ilib \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./\<$F\[@&#8203;a="b"\.\.$&\]>/; print;
> print"a=@&#8203;a"'
> \[\.\.\.\]
> SV = PVMG\(0x854ab74\) at 0x855c02c
>   REFCNT = 1
>   FLAGS = \(GMG\,SMG\)
>   IV = 0
>   NV = 0
>   PV = 0
>   MAGIC = 0x855d3f4
>     MG\_VIRTUAL = &PL\_vtbl\_sv
>     MG\_TYPE = PERL\_MAGIC\_sv\(\\0\)
>     MG\_OBJ = 0x855c01c
>     MG\_LEN = 1
>     MG\_PTR = 0x855d314 "&"
> Use of uninitialized value $& in range \(or flop\) at \-e line 1\.
> \<A>
> a=
> 
> 
> I don't know the first thing about magic or the regex engine\, so I'm
> stopping here\. Could someone take a peek?

I seem to remember looking into this before\.  IIRC\, it has to do with
s/// guessing whether it needs to set up match information for the rhs
based on its op tree\.  Or something like that\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Apr 29 11&#8203;:00&#8203;:17 2012\, sprout wrote&#8203;:
> On Sun Apr 29 09&#8203;:45&#8203;:23 2012\, Hugmeir wrote&#8203;:
> > On Tue Feb 24 14&#8203;:12&#8203;:49 2004\, rjk\-perl\-p5p@&#8203;tamias\.net wrote&#8203;:
> > > On Tue\, Feb 24\, 2004 at 11&#8203;:56&#8203;:51AM \-0500\, Casey West wrote&#8203;:
> > > > It was Monday\, February 23\, 2004 when perl\-5\.8\.0@&#8203;ton\.iguana\.be \(via
> > > RT\) took the soap box\, saying&#8203;:
> > > > : perl \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./\<$F\[@&#8203;a="b"\.\.$&\]>/; print;
> > > print"a=@&#8203;a"'
> > > > : Argument "b" isn't numeric in range \(or flop\) at \-e line 1\.
> > > > : Use of uninitialized value in range \(or flop\) at \-e line 1\.
> > > > : Use of uninitialized value in range \(or flop\) at \-e line 1\.
> > > > : \<B>
> > > > : a=0
> > > >
> > > > If I add a /e modifier on the end of your s///\, and fix up the
> > > > replacement section\, I get the following&#8203;:
> > > >
> > > > perl \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./join $F\[@&#8203;a="b"\.\.$&\]\, "\<"\,
> > > ">"/e; print; print"a=@&#8203;a"'
> > > > \<Z>
> > > > a=b c d e f g h i j k l m n o p q r s t u v w x y z
> > > >
> > > > : I expected $& to be "a"\, so @&#8203;a should become \(a b c \.\.\.\.z\)\, so
> > > fetch
> > > > : element 26 from @&#8203;F\, which should be "Z"
> > > >
> > > > I think that I got what you expected\. When you want to evaluate code
> > > > in a regular expression you must add the /e modifier\. I think that
> > > > this is notabug\. Am I missing something?
> > > 
> > > Perl evaluates code inside array and hash indexes inside double\-
> > > quotish
> > > strings\.  MJD takes advantage of this with his Interpolation module
> > > \(e\.g\. print "3 \+ 4 = $E\{3\+4\}";\)\.  The warnings from Ton's example
> > > indicate
> > > that the flip flop operator is being executed; the bug is that $&
> > > isn't
> > > being set\.
> > > 
> > > Ronald
> > > 
> > 
> > This is still present in blead\, albeit behaving a tad differently&#8203;:
> > $ \./perl \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./\<$F\[@&#8203;a="b"\.\.$&\]>/; print;
> > print"a=@&#8203;a"'
> > Use of uninitialized value $& in range \(or flop\) at \-e line 1\.
> > \<A>
> > a=
> > 
> > Strangely enough\, you can get $& to work in the flipflop if it's
> > evaluated beforehand somehow&#8203;:
> > $ \./perl \-Ilib \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./\<$F\[\!\!$& &&
> > \(@&#8203;a="b"\.\.$&\)\]>/; print; print"a=@&#8203;a"'
> > \<Z>
> > a=b c d e f g h i j k l m n o p q r s t u v w x y z
> > 
> > Putting an sv\_dump\(\) after the SvGETMAGIC\(right\) in pp\_flop gives us
this&#8203;:
> > $ \./perl \-Ilib \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./\<$F\[@&#8203;a="b"\.\.$&\]>/; print;
> > print"a=@&#8203;a"'
> > \[\.\.\.\]
> > SV = PVMG\(0x854ab74\) at 0x855c02c
> >   REFCNT = 1
> >   FLAGS = \(GMG\,SMG\)
> >   IV = 0
> >   NV = 0
> >   PV = 0
> >   MAGIC = 0x855d3f4
> >     MG\_VIRTUAL = &PL\_vtbl\_sv
> >     MG\_TYPE = PERL\_MAGIC\_sv\(\\0\)
> >     MG\_OBJ = 0x855c01c
> >     MG\_LEN = 1
> >     MG\_PTR = 0x855d314 "&"
> > Use of uninitialized value $& in range \(or flop\) at \-e line 1\.
> > \<A>
> > a=
> > 
> > 
> > I don't know the first thing about magic or the regex engine\, so I'm
> > stopping here\. Could someone take a peek?
> 
> I seem to remember looking into this before\.  IIRC\, it has to do with
> s/// guessing whether it needs to set up match information for the rhs
> based on its op tree\.  Or something like that\.

This is basically the same as \#49190\.  op\.c&#8203;:pm\_runtime\, in its if \(repl\)
block\, guesses whether the rhs of a substitution can change from one
iteration to the next\.  If it can change\, it gets a substcont op\. 
Otherwise\, pp\_subst hangs on to the same string\.  The logic in
pm\_runtime is deserving of a whole list of expletives\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Oct 08 22&#8203;:09&#8203;:22 2012\, sprout wrote&#8203;:
> On Sun Apr 29 11&#8203;:00&#8203;:17 2012\, sprout wrote&#8203;:
> > On Sun Apr 29 09&#8203;:45&#8203;:23 2012\, Hugmeir wrote&#8203;:
> > > On Tue Feb 24 14&#8203;:12&#8203;:49 2004\, rjk\-perl\-p5p@&#8203;tamias\.net wrote&#8203;:
> > > > On Tue\, Feb 24\, 2004 at 11&#8203;:56&#8203;:51AM \-0500\, Casey West wrote&#8203;:
> > > > > It was Monday\, February 23\, 2004 when perl\-5\.8\.0@&#8203;ton\.iguana\.be
\(via
> > > > RT\) took the soap box\, saying&#8203;:
> > > > > : perl \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./\<$F\[@&#8203;a="b"\.\.$&\]>/; print;
> > > > print"a=@&#8203;a"'
> > > > > : Argument "b" isn't numeric in range \(or flop\) at \-e line 1\.
> > > > > : Use of uninitialized value in range \(or flop\) at \-e line 1\.
> > > > > : Use of uninitialized value in range \(or flop\) at \-e line 1\.
> > > > > : \<B>
> > > > > : a=0
> > > > >
> > > > > If I add a /e modifier on the end of your s///\, and fix up the
> > > > > replacement section\, I get the following&#8203;:
> > > > >
> > > > > perl \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./join $F\[@&#8203;a="b"\.\.$&\]\, "\<"\,
> > > > ">"/e; print; print"a=@&#8203;a"'
> > > > > \<Z>
> > > > > a=b c d e f g h i j k l m n o p q r s t u v w x y z

> This is basically the same as \#49190\.  op\.c&#8203;:pm\_runtime\, in its if \(repl\)
> block\, guesses whether the rhs of a substitution can change from one
> iteration to the next\.  If it can change\, it gets a substcont op\. 
> Otherwise\, pp\_subst hangs on to the same string\.  The logic in
> pm\_runtime is deserving of a whole list of expletives\.
> 

In particular\, it is following op\_next pointers to look for ‘dangerous’
ops \(i\.e\.\, ops with side effects\)\.  So any logop can fool it\, as its ops
are not in a simple op\_next loop\.  Range ops are logops\.


\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Oct 08 22&#8203;:37&#8203;:14 2012\, sprout wrote&#8203;:
> On Mon Oct 08 22&#8203;:09&#8203;:22 2012\, sprout wrote&#8203;:
> > On Sun Apr 29 11&#8203;:00&#8203;:17 2012\, sprout wrote&#8203;:
> > > On Sun Apr 29 09&#8203;:45&#8203;:23 2012\, Hugmeir wrote&#8203;:
> > > > On Tue Feb 24 14&#8203;:12&#8203;:49 2004\, rjk\-perl\-p5p@&#8203;tamias\.net wrote&#8203;:
> > > > > On Tue\, Feb 24\, 2004 at 11&#8203;:56&#8203;:51AM \-0500\, Casey West wrote&#8203;:
> > > > > > It was Monday\, February 23\, 2004 when perl\-5\.8\.0@&#8203;ton\.iguana\.be
> \(via
> > > > > RT\) took the soap box\, saying&#8203;:
> > > > > > : perl \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./\<$F\[@&#8203;a="b"\.\.$&\]>/; print;
> > > > > print"a=@&#8203;a"'
> > > > > > : Argument "b" isn't numeric in range \(or flop\) at \-e line 1\.
> > > > > > : Use of uninitialized value in range \(or flop\) at \-e line 1\.
> > > > > > : Use of uninitialized value in range \(or flop\) at \-e line 1\.
> > > > > > : \<B>
> > > > > > : a=0
> > > > > >
> > > > > > If I add a /e modifier on the end of your s///\, and fix up the
> > > > > > replacement section\, I get the following&#8203;:
> > > > > >
> > > > > > perl \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./join $F\[@&#8203;a="b"\.\.$&\]\, "\<"\,
> > > > > ">"/e; print; print"a=@&#8203;a"'
> > > > > > \<Z>
> > > > > > a=b c d e f g h i j k l m n o p q r s t u v w x y z
> 
> > This is basically the same as \#49190\.  op\.c&#8203;:pm\_runtime\, in its if \(repl\)
> > block\, guesses whether the rhs of a substitution can change from one
> > iteration to the next\.  If it can change\, it gets a substcont op\. 
> > Otherwise\, pp\_subst hangs on to the same string\.  The logic in
> > pm\_runtime is deserving of a whole list of expletives\.
> > 
> 
> In particular\, it is following op\_next pointers to look for ‘dangerous’
> ops \(i\.e\.\, ops with side effects\)\.  So any logop can fool it\, as its ops
> are not in a simple op\_next loop\.  Range ops are logops\.

I have fixed this particular case in commit 86e69b18632c\, by foregoing
the optimisation when the replacement contains logops\.  I am marking
this as resolved\, as bug \#49190 lists the other problems\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @rurban

This was not a resolving fix\.

You simply disable s/// optimizations when a LOGOP is present\. 

You correctly analyzed it\, but you should follow the LOGOP 
alternative path instead if disabling the optimization at all\.

On Tue Oct 09 17&#8203;:52&#8203;:00 2012\, sprout wrote&#8203;:
> On Mon Oct 08 22&#8203;:37&#8203;:14 2012\, sprout wrote&#8203;:
> > On Mon Oct 08 22&#8203;:09&#8203;:22 2012\, sprout wrote&#8203;:
> > > On Sun Apr 29 11&#8203;:00&#8203;:17 2012\, sprout wrote&#8203;:
> > > > On Sun Apr 29 09&#8203;:45&#8203;:23 2012\, Hugmeir wrote&#8203;:
> > > > > On Tue Feb 24 14&#8203;:12&#8203;:49 2004\, rjk\-perl\-p5p@&#8203;tamias\.net wrote&#8203;:
> > > > > > On Tue\, Feb 24\, 2004 at 11&#8203;:56&#8203;:51AM \-0500\, Casey West wrote&#8203;:
> > > > > > > It was Monday\, February 23\, 2004 when perl\-
5\.8\.0@&#8203;ton\.iguana\.be
> > \(via
> > > > > > RT\) took the soap box\, saying&#8203;:
> > > > > > > : perl \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./\<$F\[@&#8203;a="b"\.\.$&\]>/; 
print;
> > > > > > print"a=@&#8203;a"'
> > > > > > > : Argument "b" isn't numeric in range \(or flop\) at \-e line 
1\.
> > > > > > > : Use of uninitialized value in range \(or flop\) at \-e line 
1\.
> > > > > > > : Use of uninitialized value in range \(or flop\) at \-e line 
1\.
> > > > > > > : \<B>
> > > > > > > : a=0
> > > > > > >
> > > > > > > If I add a /e modifier on the end of your s///\, and fix up 
the
> > > > > > > replacement section\, I get the following&#8203;:
> > > > > > >
> > > > > > > perl \-wle '@&#8203;F="A"\.\."BZ"; $\_="a"; s/\./join $F\[@&#8203;a="b"\.\.$&\]\, 
"\<"\,
> > > > > > ">"/e; print; print"a=@&#8203;a"'
> > > > > > > \<Z>
> > > > > > > a=b c d e f g h i j k l m n o p q r s t u v w x y z
> > 
> > > This is basically the same as \#49190\.  op\.c&#8203;:pm\_runtime\, in its if 
\(repl\)
> > > block\, guesses whether the rhs of a substitution can change from 
one
> > > iteration to the next\.  If it can change\, it gets a substcont op\. 
> > > Otherwise\, pp\_subst hangs on to the same string\.  The logic in
> > > pm\_runtime is deserving of a whole list of expletives\.
> > > 
> > 
> > In particular\, it is following op\_next pointers to look for 
‘dangerous’
> > ops \(i\.e\.\, ops with side effects\)\.  So any logop can fool it\, as its 
ops
> > are not in a simple op\_next loop\.  Range ops are logops\.
> 
> I have fixed this particular case in commit 86e69b18632c\, by foregoing
> the optimisation when the replacement contains logops\.  I am marking
> this as resolved\, as bug \#49190 lists the other problems\.

\-\- 
Reini Urban
